python/tank/bootstrap/threading.py

```diff
@@ -0,0 +1,269 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+try:
+    # Import Qt without having to worry about the version to use.
+    from ..util.qt_importer import QtImporter
+    importer = QtImporter()
+    QtCore = importer.QtCore
+    QtGui = importer.QtGui
+except ImportError:
+    # Raise an exception when Qt is not available.
+    raise
+
+from ..api import Sgtk
+
+
+class BootstrapSupervisor(QtCore.QObject):
```

Comment:
The `BootstrapSupervisor` class needs to subclass from `QObject` since it has some slot methods that call `self.thread()` to make the worker operate with the main thread affinity where the main event loop can handle its deletion.